Providers/Scripts/2.4x-2.5x/Scripts/nxOMSAgentNPMConfig.py

```diff
@@ -163,19 +254,122 @@ def Test(ConfigType, ConfigID, Contents, Ensure, ContentChecksum):
         #compare
         if origConfigData is None or origConfigData != Contents:
             retval = -1
-    return [retval]
+    return retval
 
+def SetConfigUpdate(Contents):
+    destFileFullPath = CONFIG_PATH.__add__(DEST_FILE_NAME)
 
-def Get(ConfigType, ConfigID, Contents, Ensure, ContentChecksum):
-    retval = 0
-    ShowMof('GET', ConfigType, ConfigID, Contents, Ensure, ContentChecksum)
-    return [retval]
+    # Update config after checking if directory exists
+    if not os.path.exists(CONFIG_PATH):
+        LG().Log('ERROR', 'CONFIG_PATH does not exist')
+        retval = -1
+    else:
+        retval = WriteFile(destFileFullPath, Contents)
+        if retval == 0:
+            LG().Log('INFO', 'Updated the file, going to notify server')
+            retval = NotifyServer(Commands.Config)
+    return retval
 
+def SetFilesUpdate(newVersion):
+    retval = UpdateAgentBinary(newVersion)
+    retval &= UpdatePluginFiles()
+    if retval:
+        return 0
+    return -1
+
+def UpdateAgentBinary(newVersion):
+    retval = True
+    arch = platform.architecture()
+    src = ''
+    if arch is not None and arch[0] == X64:
+        src = RESOURCE_MODULE_PATH.__add__(DSC_X64_AGENT_PATH)
+        retval &= DeleteAllFiles(src, AGENT_BINARY_PATH)
+        retval &= CopyAllFiles(src, AGENT_BINARY_PATH)
+    else:
+        src = RESOURCE_MODULE_PATH.__add__(DSC_X86_AGENT_PATH)
+        retval &= DeleteAllFiles(src, AGENT_BINARY_PATH)
+        retval &= CopyAllFiles(src, AGENT_BINARY_PATH)
+
+    # set capabilities to binary
+    src_files = os.listdir(AGENT_BINARY_PATH)
+    for file_name in src_files:
+        full_file_name = os.path.join(src, file_name) # assuming only file in directory
+        break
+    NPM_ACTION.binary_setcap(full_file_name)
+
+    #TODO: Notify ruby plugin
+    #retval &= NotifyServer(Commands.RestartNPM)
+
+    #Update version number
+    WriteFile(AGENT_RESOURCE_VERSION_PATH, newVersion)
+    return retval
 
-def Print(s, file=sys.stdout):
-    file.write(s + '\n')
+def UpdatePluginFiles():
+    retval = True
+
+    #stop oms agent
+    if not OMS_ACTION.stop_oms_agent():
+        return False
+
+    #replace files
+    retval &= DeleteAllFiles(RESOURCE_MODULE_PATH.__add__(DSC_PLUGIN_PATH), PLUGIN_PATH)
+    retval &= DeleteAllFiles(RESOURCE_MODULE_PATH.__add__(DSC_PLUGIN_CONF_PATH), PLUGIN_CONF_PATH)
+    retval &= CopyAllFiles(RESOURCE_MODULE_PATH.__add__(DSC_PLUGIN_PATH), PLUGIN_PATH)
+    retval &= CopyAllFiles(RESOURCE_MODULE_PATH.__add__(DSC_PLUGIN_CONF_PATH), PLUGIN_CONF_PATH)
+
+    # start oms agent
+    retval &= OMS_ACTION.start_oms_agent()
```

Comment:
currently, the DSC is triggered by fluentd. i'm afraid it might cause issue if there is a gap between stop and start by chance.
the process won't lock the files. you can call service_control restart after copying the files.